Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -201,252 +82,618 @@ void LoadSettings()
+						// skip loading renamed/deleted docked ships
+						if(!accData){
+							continue;
+						}
+
+						nameToDockedInfoMap[dockedShipName].carrierName = carrierName.c_str();
+						nameToDockedInfoMap[dockedShipName].lastDockedSolar = CreateID(ini.get_value_string(1));
+
+						ci.dockedShipList.push_back(dockedShipName);
+						dockedCount++;
+					}
+				}
+				if (doLoad) {
+					nameToCarrierInfoMap[carrierName] = ci;
+					carrierCount++;
+				}
+				else
+				{
+					for (const auto& dockedShipName : ci.dockedShipList)
+					{
+						MoveOfflineShipToLastDockedSolar(dockedShipName);
+					}
+				}
+			}
+		}
+		ini.close();
+	}
 	ConPrint(L"DockingModules: Loaded %u equipment\n", dockingModAmount);
+	ConPrint(L"DockingModules: Found %u ships docked on %u carriers\n", dockedCount, carrierCount);
 	ConPrint(L"DockingModules: Allowing ships below the cargo capacity of %i to dock\n", cargoCapacityLimit);
 }
 
-void __stdcall BaseExit(uint iBaseID, uint iClientID)
+void SaveData()
 {
-	returncode = DEFAULT_RETURNCODE;
+	if (nameToCarrierInfoMap.empty()){
+		return;
+	}
+	char datapath[MAX_PATH];
+	GetUserDataPath(datapath);
+	string path = string(datapath) + "\\Accts\\MultiPlayer\\docking_module\\mobile_docking.ini";
+	FILE *file = fopen(path.c_str(), "w");
+	if (file)
+	{
+		for (const auto& ci : nameToCarrierInfoMap)
+		{
+			if (ci.second.dockedShipList.empty()) {
+				continue;
+			}
+			fprintf(file, "[CarrierData]\n");
+			fprintf(file, "carrier = %ls\n", ci.first.c_str());
+			fprintf(file, "lastLogin = %u\n", (uint)ci.second.lastCarrierLogin);
+			for (const wstring& dockedName : ci.second.dockedShipList)
+			{
+				const auto& lastSolarInfo = Universe::get_base(nameToDockedInfoMap[dockedName.c_str()].lastDockedSolar);
+				fprintf(file, "docked = %ls, %s\n", dockedName.c_str(), lastSolarInfo->cNickname);
+			}
+		}
 
-	mobiledockClients[iClientID].iDockingModulesInstalled = GetInstalledModules(iClientID);
+		fclose(file);
+	}
+}
+
+void dockShipOnCarrier(uint dockingID, uint carrierID)
+{
 
-	// Normalize the docking modules available, with the number of people currently docked
-	mobiledockClients[iClientID].iDockingModulesAvailable = (mobiledockClients[iClientID].iDockingModulesInstalled - mobiledockClients[iClientID].mapDockedShips.size());
+	// The client is free to dock, erase from the pending list and handle
+	mapPendingDockingRequests.erase(dockingID);
 
-	// If this is a ship which is currently docked, clear the market
-	if (mobiledockClients[iClientID].mobileDocked)
+	if (mobiledockClients[carrierID].iDockingModulesAvailable == 0)
 	{
-		SendResetMarketOverride(iClientID);
+		PrintUserCmdText(dockingID, L"Carrier has no free docking capacity");
+		return;
 	}
 
-}
-
-// Temporary storage for client data to be handled in LaunchPosHook.
-CLIENT_DATA undockingShip;
+	string scProxyBase = HkGetPlayerSystemS(carrierID) + "_proxy_base";
+	uint iBaseID;
+	Universe::IBase* baseInfo;
+	if (targetMobileDockingBase)

```

Comment:
The variable names are becoming confusing. I would suggest a comment to defer the different types of bases.